### PR TITLE
[FIX] website: allow user to edit theme view

### PR DESCRIPTION
**Pull request by ngochuy97hp**
**Current behavior before PR:**

Currently, "Editor and Designer" can't edit the views which originate
from a theme (which are copies from the theme `theme.ir.ui.view`
records).

This error is caused by https://github.com/odoo/odoo/commit/1eb7c577ecb31c2f0876b6dd76769e2b47802cb0, which is preventing the `arch_updated`
field to be set to `True`, as this field is supposed to be set to `True`
only on manual user change, not during module update.

Indeed, restricted users don't have access to the themes records, making
the overide raising an error.

But there were actually no reason for this code to be executed outside
module operation.

Steps to reproduce:
- Create any view in a theme module.
  For example: create a custom footer view.
- As a "Editor and Designer" user, edit this view through the website
  builder.
- An access right error is raised.

**Desired behavior after PR is merged:** 
Allow user to edit theme view

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by ngochuy97hp** [comment]
@rdeodoo Could you review this PR?

File: addons/website/models/theme_models.py
Comment (rdeodoo) [inline addons/website/models/theme_models.py:282]: I think this is fixing the consequence and not the root cause.
https://github.com/odoo/odoo/commit/1eb7c577ecb31c2f0876b6dd76769e2b47802cb0 is probably missing a `if` condition on top of that whole code to check if we are in a module update state, that's the only flow where we want to perform that check.

**Review by rdeodoo — CHANGES_REQUESTED** [review]
Thank you for your contribution.

File: addons/website/models/theme_models.py
Comment (ngochuy97hp) [reply addons/website/models/theme_models.py:282]: @rdeodoo Thank you. It means that when i upgrade the theme, it will execute this code right? I still haven't found a sign yet to know if it's in an upgrading state the module or not while in View. Or maybe I have to add some context somewhere else? Need your more explanation.

Comment (rdeodoo) [reply addons/website/models/theme_models.py:282]: Sorry for the late reply, you can check with `if self.pool._init:`
Around this https://github.com/odoo/odoo/commit/1eb7c577ecb31c2f0876b6dd76769e2b47802cb0#diff-4e6ab140edc4e1362791f5184ab01d23b5a27969011a27f998c9ef804dd64d53R225
That should fix the root cause

Comment (ngochuy97hp) [reply addons/website/models/theme_models.py:285]: Done. That's so great! It was a new knowledge for me. Thank you very much! :smile:

File: addons/website/models/theme_models.py
Comment (rdeodoo) [inline addons/website/models/theme_models.py:280]:
```suggestion
            return super().write(vals)
```
Python 3 :+1:

**Review by rdeodoo — CHANGES_REQUESTED** [review]
Thanks again for your contribution.

Since you seems interested by our guidelines, if you don't mind some more, here is 3 small nitpicking ones, tell me if you want to ignore it and I'll have a final functional check and r+ this one:
- Python3 super
- You should use all the space available up to 72 char / line in your commit message, not randomly returning at 45 :)
- Use markdown link

So
```
Currently, user with "Editor and Designer" right
cannot edit views that created from the theme module.
This error caused by https://github.com/odoo/odoo/commit/1eb7c577ecb31c2f0876b6dd76769e2b47802cb0. We now correct
it by adding condition on top of that whole code to
check if we are in a module update state.

Steps to reproduce:
- Create any view in the theme module.
  For example: create a custom view footer.
- User with "Editor and Designer" right edit
  this view by using website editor.
```

=> 

```
Currently, user with "Editor and Designer" right cannot edit views that
created from the theme module.
This error is caused by [1].
We now correct it by adding condition on top of that whole code to check
if we are in a module update state, as that code is only supposed to be
run in that state.

Steps to reproduce:
- Create any view in the theme module.
  For example: create a custom view footer.
- User with "Editor and Designer" right edit this view by using website
  editor.

[1]: https://github.com/odoo/odoo/commit/1eb7c577ecb31c2f0876b6dd76769e2b47802cb0
```
(I just also added `as that code is only supposed to be run in that state.`)

File: addons/website/models/theme_models.py
Comment (rdeodoo) [reply addons/website/models/theme_models.py]: You are welcome

File: addons/website/models/theme_models.py
Comment (ngochuy97hp) [reply addons/website/models/theme_models.py:280]: Thank you. I just wrote it to look like the code below, no other reason. I will fix it.

**Comment by ngochuy97hp** [comment]
> 

I agree, I will edit according to your rule, it's worth learning :)
... but unfortunately the test run failed :(

**Comment by ngochuy97hp** [comment]
> Thanks again for your contribution.
> 
> Since you seems interested by our guidelines, if you don't mind some more, here is 3 small nitpicking ones, tell me if you want to ignore it and I'll have a final functional check and r+ this one:
> 
> * Python3 super
> * You should use all the space available up to 72 char / line in your commit message, not randomly returning at 45 :)
> * Use markdown link
> 
> So
> 
> ```
> Currently, user with "Editor and Designer" right
> cannot edit views that created from the theme module.
> This error caused by https://github.com/odoo/odoo/commit/1eb7c577ecb31c2f0876b6dd76769e2b47802cb0. We now correct
> it by adding condition on top of that whole code to
> check if we are in a module update state.
> 
> Steps to reproduce:
> - Create any view in the theme module.
>   For example: create a custom view footer.
> - User with "Editor and Designer" right edit
>   this view by using website editor.
> ```
> 
> =>
> 
> ```
> Currently, user with "Editor and Designer" right cannot edit views that
> created from the theme module.
> This error is caused by [1].
> We now correct it by adding condition on top of that whole code to check
> if we are in a module update state, as that code is only supposed to be
> run in that state.
> 
> Steps to reproduce:
> - Create any view in the theme module.
>   For example: create a custom view footer.
> - User with "Editor and Designer" right edit this view by using website
>   editor.
> 
> [1]: https://github.com/odoo/odoo/commit/1eb7c577ecb31c2f0876b6dd76769e2b47802cb0
> ```
> 
> (I just also added `as that code is only supposed to be run in that state.`)

I agree, I will edit according to your rule, it's worth learning :)
... but unfortunately the test run failed :(

File: addons/website/models/theme_models.py
Comment (rdeodoo) [inline addons/website/models/theme_models.py:279]: This is a bit tricky as this is related to the tests, which tries to simulate the best they can the flow they are testing.

In this case it is because https://github.com/odoo/odoo/commit/1eb7c577ecb31c2f0876b6dd76769e2b47802cb0#diff-86206ee83b8963a647f6373486c147e25cf21204a3fc812368d3dad587c5dd99R1316 is simulating a module update, which in real case is going through this `write()` as `pool._init` is True.
But in the test, it is not True, as this is not a real update (just simulating).

The workaround would be to do
```
import threading
# [...]
test_mode = getattr(threading.currentThread(), 'testing', False)
if not (test_mode or self.pool._init):
```
That's not exactly correct as the tests will now avoid this condition, acting as it was doing an update, but that's not that bad IMHO, as it was like that without your PR already.

The other solution might be to do in the test line I mentioned `with_context(simulate_module_update=True)` and here check if we have that key context. As you prefer.

I am not a big fan of writing code related to tests in our regular codebase, but in this case I don't see a better way, if we want to keep the test.

Let me know if you prefer that I finish this PR, both ways are ok to me

Comment (rdeodoo) [reply addons/website/models/theme_models.py:279]: @xmo-odoo May I ask your opinion here about what I said?
Is there a better way you can think of?

**Comment by ngochuy97hp** [comment]
@rdeodoo I'm a bit wondering, why does using ``getattr`` contain security risk?
https://runbot.odoo.com/runbot/build/14900248

**Comment by xmo-odoo** [comment]
>  I'm a bit wondering, why does using `getattr` contain security risk? [runbot.odoo.com/runbot/build/14900248](https://runbot.odoo.com/runbot/build/14900248)

Sometimes changes are submitted which use `getattr` with the attribute name under attacker control, leading to possible unexpected information disclosure.

The security flag thing is not very smart (it's just a bunch of regex) so it's not able to exclude cases with a static attribute name, or whitelist specific patterns (like this one).

File: addons/website/models/theme_models.py
Comment (xmo-odoo) [reply addons/website/models/theme_models.py]: > May I ask your opinion here about what I said? Is there a better way you can think of?

Not really, I can't think of something significantly better.

The code comment is pretty unhelpful though, it says what the code already says which is redundant, *why* this code is the way it is would be more useful aka the purpose of the override, and of bypassing it.

I think the issue caused by the override would be useful too, "there's an error" is not really helpful to the reader of the commit message, what the error is and why it happens would be more useful: it looks like the override tries to be careful when performing comparisons, so one could expect regular editions to work fine (I assume that is what happened originall).

Also while at it, the return values of the sub-writes in the overide are useles (and uselessly combined): `write` should just return a static `True`. In case of errors, it just raises an exception.

Comment (ngochuy97hp) [reply addons/website/models/theme_models.py]: Yes, i will change it. The new message is quite easy to understand. Sorry for my limited English, so the commit message is not detailed.

**Comment by ngochuy97hp** [comment]
> > I'm a bit wondering, why does using `getattr` contain security risk? [runbot.odoo.com/runbot/build/14900248](https://runbot.odoo.com/runbot/build/14900248)
> 
> Sometimes changes are submitted which use `getattr` with the attribute name under attacker control, leading to possible unexpected information disclosure.
> 
> The security flag thing is not very smart (it's just a bunch of regex) so it's not able to exclude cases with a static attribute name, or whitelist specific patterns (like this one).

Thank you!

**Comment by ngochuy97hp** [comment]
@rdeodoo @xmo-odoo Could you guys help me, please? What do you guys think? What should i do next?

**Comment by rdeodoo** [comment]
Just adapt the code comment line as stated here https://github.com/odoo/odoo/pull/88990#discussion_r861071551

Just replace it with
```py
# During a theme module update, theme views' copies receiving an arch
# update should not be considered as `arch_updated`, as this is not a
# user made change.
```
and that should be it, XMO requests should be processed.

**Comment by rdeodoo** [comment]
@xmo-odoo Do you mind approving the security CI here?

**Review by rdeodoo — APPROVED** [review]
Waiting for the security CI approval before shipping this one.

Note that arguably, now that we introduced some test-only code in the codebase, we could have used a context key only used by the tests to enable the `arch_updated` behavior instead of checking if we are in test mode.
That would have been a bit more flexible as a test could have tested both ways, but that would require a new context key.

**Comment by ngochuy97hp** [comment]
> Waiting for the security CI approval before shipping this one.
> 
> Note that arguably, now that we introduced some test-only code in the codebase, we could have used a context key only used by the tests to enable the `arch_updated` behavior instead of checking if we are in test mode. That would have been a bit more flexible as a test could have tested both ways, but that would require a new context key.

Thank you!

**Comment by xmo-odoo** [comment]
@robodoo override=ci/security

Test mode check.

**Comment by rdeodoo** [comment]
@robodoo r+

Thank you for your contribution

**Pull request closed without merging**
